src/mruby.rs

```diff
@@ -0,0 +1,439 @@
+// mrusty. mruby bindings for Rust
+// Copyright (C) 2016  Dragoș Tiselice
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::mem;
+use std::rc::Rc;
+
+use super::mruby_ffi::*;
+
+/// A safe `struct` for the mruby API.
+///
+/// # Examples
+///
+/// ```
+/// # use mrusty::MRuby;
+/// # use mrusty::run;
+/// let mruby = MRuby::new();
+/// let result = run(&mruby, "2 + 2 == 5").unwrap();
+///
+/// assert_eq!(result.to_bool().unwrap(), false);
+/// ```
+pub struct MRuby {
+    mrb: *mut MRState,
+    ctx: *mut MRContext,
+    classes: Box<HashMap<String, (*mut MRClass, MRDataType)>>,
+    methods: Box<HashMap<String, Box<HashMap<u32, Box<Fn(&Rc<RefCell<MRuby>>, Value) -> Value>>>>>
+}
+
+impl MRuby {
+    /// Creates an mruby state and context stored in a `struct`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use mrusty::MRuby;
+    /// let mruby = MRuby::new();
+    /// ```
+    pub fn new() -> Rc<RefCell<MRuby>> {
+        unsafe {
+            let mrb = mrb_open();
+
+            let mruby = Rc::new(RefCell::new(
+                MRuby {
+                    mrb: mrb,
+                    ctx: mrbc_context_new(mrb),
+                    classes: Box::new(HashMap::new()),
+                    methods: Box::new(HashMap::new())
+                }
+            ));
+
+            mrb_ext_set_ud(mrb, mem::transmute::<&Rc<RefCell<MRuby>>, *const u8>(&mruby));
```

Comment:
I think this is the source of rust-lang/rust#31790. `&mruby` is a pointer to an `Rc`, but that pointer will be invalid once this method returns.